File: src/BusinessLogic/Handlers/CallbackQueryHandlers/InfoCallbackQueryHandler.cs
using Bounan.Bot.BusinessLogic.CommandDto;
using Bounan.Bot.BusinessLogic.Handlers.MessageHandlers;
using Bounan.Bot.TelegramBot.Telegram;
using Microsoft.Extensions.Logging;
using Telegram.Bot.Requests;
using Telegram.Bot.Types;

namespace Bounan.Bot.BusinessLogic.Handlers.CallbackQueryHandlers;

public class InfoCallbackQueryHandler(
    ILogger<InfoCallbackQueryHandler> logger,
    InfoMessageHandler infoMessageHandler) : ICallbackQueryHandler
{
    public static bool CanHandle(CallbackQuery item) => item.Data?.StartsWith(InfoCommandDto.Command) ?? false;

    public async Task<AnswerCallbackQueryRequest> HandleAsync(
        CallbackQuery updateItem,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(updateItem);
        ArgumentNullException.ThrowIfNull(updateItem.Data);
        ArgumentNullException.ThrowIfNull(updateItem.Message);
        ArgumentNullException.ThrowIfNull(updateItem.Message.Chat);

        logger.LogInformation("Handling info callback query {Data}", updateItem.Data);

        await infoMessageHandler.HandleAsync(
            new Message
            {
                Chat = updateItem.Message.Chat,
                Text = updateItem.Data,
            },
            cancellationToken);

        return new AnswerCallbackQueryRequest(updateItem.Id);
    }
}

File: src/handlers/on-webhook/event-handlers/callback-query-adapters/info-callback-query-handler.ts
import { CallbackQuery } from '@lightweight-clients/telegram-bot-api-lightweight-client';

import { InfoCommandDto } from '../../command-dtos';
import { infoMessageHandler } from '../message-handlers/info-message-handler';
import { CallbackQueryHandler } from '../query-handler';
import { processCallbackQuery } from './callback-query-adapter-base';

const canHandle = (callbackQuery: CallbackQuery) => callbackQuery.data?.startsWith(InfoCommandDto.Command) || false;

const handler: CallbackQueryHandler = async (callbackQuery) =>
    await processCallbackQuery(callbackQuery, infoMessageHandler.handler);

export const infoCallbackQueryHandler = {
    canHandle,
    handler,
}